File: test/benchmark/beng/js/revcomp.js
// BENG Benchmark: reverse-complement (Node.js reference)
// Read FASTA, output reverse-complement of each sequence

const fs = require('fs');

const INPUT_PATH = process.argv[2] || "test/benchmark/beng/input/fasta_1000.txt";
const LINE_WIDTH = 60;

const COMPLEMENT = {
    'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C',
    'M': 'K', 'K': 'M', 'R': 'Y', 'Y': 'R',
    'V': 'B', 'B': 'V', 'H': 'D', 'D': 'H',
    'W': 'W', 'S': 'S', 'N': 'N',
    'a': 'T', 't': 'A', 'c': 'G', 'g': 'C',
    'm': 'K', 'k': 'M', 'r': 'Y', 'y': 'R',
    'v': 'B', 'b': 'V', 'h': 'D', 'd': 'H',
    'w': 'W', 's': 'S', 'n': 'N'
};

function outputReverseComplement(header, seq) {
    console.log(header);
    const upper = seq.toUpperCase();
    let comp = '';
    for (let i = 0; i < upper.length; i++) {
        comp += COMPLEMENT[upper[i]] || upper[i];
    }
    const rev = comp.split('').reverse().join('');
    for (let pos = 0; pos < rev.length; pos += LINE_WIDTH) {
        console.log(rev.substring(pos, pos + LINE_WIDTH));
    }
}

const text = fs.readFileSync(INPUT_PATH, 'utf-8');
const lines = text.split('\n');

let header = '';
let seq = '';

for (const line of lines) {
    if (line.length > 0 && line[0] === '>') {
        if (seq.length > 0) {
            outputReverseComplement(header, seq);
        }
        header = line;
        seq = '';
    } else if (line.length > 0) {
        seq += line.toUpperCase();
    }
}

if (seq.length > 0) {
    outputReverseComplement(header, seq);
}

File: test/benchmark/beng/python/revcomp.py
#!/usr/bin/env python3
# BENG Benchmark: reverse-complement (Python)
# Read FASTA, output reverse-complement of each sequence
import sys
import time

INPUT_PATH = sys.argv[1] if len(sys.argv) > 1 else "test/benchmark/beng/input/fasta_1000.txt"
LINE_WIDTH = 60

COMPLEMENT = str.maketrans(
    'ACTGMKRYVBHDWSNacgtmkryvbhdwsn',
    'TGACKMYRBVDHWSNTGACkMYRBVDHWSN'
)

with open(INPUT_PATH, 'r') as f:
    text = f.read()

lines = text.split('\n')

t0 = time.perf_counter_ns()
header = ''
seq = []
for line in lines:
    if line and line[0] == '>':
        if seq:
            combined = ''.join(seq).upper()
            rev_comp = combined.translate(COMPLEMENT)[::-1]
            print(header)
            for pos in range(0, len(rev_comp), LINE_WIDTH):
                print(rev_comp[pos:pos + LINE_WIDTH])
        header = line
        seq = []
    elif line:
        seq.append(line)

if seq:
    combined = ''.join(seq).upper()
    rev_comp = combined.translate(COMPLEMENT)[::-1]
    print(header)
    for pos in range(0, len(rev_comp), LINE_WIDTH):
        print(rev_comp[pos:pos + LINE_WIDTH])

t1 = time.perf_counter_ns()
print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")
